src/backends/perf_pt/perf_pt.c

```diff
@@ -348,124 +318,173 @@ tracer_thread(void *arg)
  */
 
 /*
- * Turn on Intel PT.
- *
- * Arguments:
- *   tr_conf: Tracer configuration.
- *
- * Returns a pointer to a tracer context, or NULL on error.
+ * Initialise a tracer context.
  */
 struct tracer_ctx *
-perf_pt_start_tracer(struct tracer_conf *tr_conf)
+perf_pt_init_tracer(struct tracer_conf *tr_conf)
 {
-    bool failing = false;
     struct tracer_ctx *tr_ctx = NULL;
-    int perf_fd = -1;
-    int clean_sem = 0, clean_thread = 0;
-    struct tracer_ctx *ret = NULL;
+    bool failing = false;
 
     // Allocate and initialise tracer context.
     tr_ctx = malloc(sizeof(*tr_ctx));
     if (tr_ctx == NULL) {
         goto clean;
     }
-    tr_ctx->trace_buf = NULL;
-    tr_ctx->trace_len = 0;
 
-    // Get the perf fd
-    perf_fd = open_perf(tr_conf->target_tid);
-    if (perf_fd == -1) {
+    // Set default values.
+    memset(tr_ctx, 0, sizeof(*tr_ctx));
+    tr_ctx->stop_fds[0] = tr_ctx->stop_fds[1] = -1;
+    tr_ctx->perf_fd = -1;
+
+    // Obtain a file descriptor through which to speak to perf.
+    tr_ctx->perf_fd = open_perf(tr_conf->target_tid);
+    if (tr_ctx->perf_fd == -1) {
+        failing = true;
+        goto clean;
+    }
+
+    // Allocate mmap(2) buffers for speaking to perf.
+    //
+    // These regions are mmap(2)ed from the perf fd in two separate chunks:
+    //
+    //            "Base buffer"
+    // -----------------------------------     ------------------
+    // | header  |       data buffer     | ... |   AUX buffer   |
+    // -----------------------------------     ------------------
+    //           ^ header->data_offset
+    //
+    // The AUX buffer is where the kernel exposes control flow packets, whereas
+    // the data buffer is used for all other kinds of packet.
+
+    // Allocate the base buffer.
+    //
+    // Data buffer is preceded by one management page (the header), hence `1 +
```

Comment:
The header mst be exactly 1 page big? What happens if the page size changes?

Reply:
Whilst the system is running? I think you would have other issues ;)

Reply:
I mean: it's a weird way of specifying a data-structure size. I'm checking that I understood it correctly!

Reply:
It is weird, I agree.

But the page size is fixed at kernel compile time, so don't worry about that.

FWIW, here's the same in the libipt docs:

Reply:
Missing text?

Reply:
yeah, hit the wrong button, updated.

Reply:
OK.